ComponentKitTests/CKComponentDelegateAttributeTests.mm

```diff
@@ -0,0 +1,123 @@
+#import "CKComponent.h"
+#import "CKComponentDelegateAttribute.h"
+#import "CKComponentViewInterface.h"
+#import "CKComponentLayout.h"
+#import "CKComponentSubclass.h"
+#import "CKComponentInternal.h"
+
+
+@interface CKDetectScrollComponent : CKCompositeComponent <UIScrollViewDelegate>
+@property (nonatomic, assign) BOOL receivedScroll;
+@end
+
+@interface CKComponentGestureActionsTests : XCTestCase
+@end
+
+@interface CKComponentDelegateAttributeTests : XCTestCase
+@end
+
+@implementation CKComponentDelegateAttributeTests
+
+- (void)testApplicationApplies
+{
+
+  CKComponentViewAttributeValue attr = CKComponentDelegateAttribute(@selector(setDelegate:), {
+    {@selector(scrollViewDidScroll:)}
+  });
+
+  UIScrollView *view = [[UIScrollView alloc] init];
+
+  attr.first.applicator(view, attr.second);
+  XCTAssertNotNil(view.delegate, @"Expected delegate to be set");
+
+  attr.first.unapplicator(view, attr.second);
+  XCTAssertNil(view.delegate, @"Expected delegate to be unset");
+}
+
+static UIScrollView *findScrollView(UIView *v) {
+  if ([v isKindOfClass:[UIScrollView class]]) {
+    return (UIScrollView *)v;
+  } else {
+    for (UIView *sub in v.subviews) {
+      return findScrollView(sub);
+    }
+  }
+  return nil;
+};
+
+
+- (void)testProxiedEventsProxy
+{
+  CKDetectScrollComponent *hierarchy =
+  [CKDetectScrollComponent
+   newWithComponent:[CKComponent
+   newWithView:{[UIScrollView class],
+     {CKComponentDelegateAttribute(@selector(setDelegate:), {
+       @selector(scrollViewDidScroll:),
+     })}}
+   size:{}]];
+
+
+  CKComponentLayout layout = [hierarchy layoutThatFits:{} parentSize:{NAN, NAN}];
+
+  UIView *container = [UIView new];
+  NSSet *mounted = CKMountComponentLayout(layout, container);
+
+  XCTAssertFalse(hierarchy.receivedScroll, @"Should not have triggered yet");
+
+  UIScrollView *scroll = findScrollView(container);
+
+  scroll.contentOffset = CGPointMake(0, 100);
+
+  XCTAssertTrue(hierarchy.receivedScroll, @"Should have recived scroll event");
+
+  // Temporary hack because there's not a good way to unmount components. Assert fails otherwise.
+  [mounted makeObjectsPerformSelector:@selector(unmount)];
+
+  CKDetectScrollComponent *noScrollHierarchy =
+  [CKDetectScrollComponent
+   newWithComponent:[CKComponent
+                     newWithView:{[UIScrollView class],
+                       {CKComponentDelegateAttribute(@selector(setDelegate:), {})}}
+                     size:{}]];
+
+  layout = [noScrollHierarchy layoutThatFits:{} parentSize:{NAN, NAN}];
+
+  CKMountComponentLayout(layout, container);
+
+  XCTAssertFalse(noScrollHierarchy.receivedScroll, @"Should not have triggered yet");
+  hierarchy.receivedScroll = NO;
+
+  scroll.contentOffset = CGPointMake(0, 100);
+
+  XCTAssertFalse(noScrollHierarchy.receivedScroll, @"Should not have triggered because we don't want scroll events.");
+  XCTAssertFalse(hierarchy.receivedScroll, @"Should not have triggered on old hierarchy either.");
+
```

Comment:
Totally. I think this test could be split up into several smaller ones:

```objc
- (void)testMountingComponentWithViewHierarchyDoesNotTriggerScrollEvent { /* ... */ }
- (void)testScrollingComponentWithViewHierarchyTriggersScrollEvent { /* ... */ }
- (void)testMountingComponentWithoutViewHierarchyDoesNotTriggerScrollEvent { /* ... */ }
- (void)testScrollingComponentWithoutViewHierarchyTriggersScrollEvent { /* ... */ }
```

From what I understand of the intent of the tests, splitting them up should make it possible to get rid of the `-makeObjectsPerformSelector:` line.